enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,384 @@
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@pmtk, MicroShift contributor"
+  - "@TurboTurtle, SOS contributor"
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-01-17
+last-updated: 2023-01-17
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-389
+see-also:
+  - "/enhancements/oc/must-gather.md"
+  - "https://sos.readthedocs.org/"
+---
+
+# MicroShift supportability tools
+
+## Summary
+This enhancement proposes the addition of supportability tooling to MicroShift.
+
+To debug issues in the cluster we need a single command that any user or
+administrator is able to run to gather all the required information to diagnose
+and solve problems.
+
+## Motivation
+Most readers will be familiar with `must-gather` for OpenShift clusters and
+what it achieves. For MicroShift we need a tool with the same principles and
+goals, which is to gather every bit of information from the cluster to be able
+to diagnose and fix issues.
+
+### User Stories
+
+* As a MicroShift administrator, I want to be able to get a report about current
+  cluster status.
+* As a MicroShift developer, I want to be able to diagnose issues with a detailed
+  debug report.
+
+### Goals
+* Provide a single tool to automate the report creation.
+
+### Non-Goals
+N/A
+
+## Proposal
+The MicroShift gathering tool is a plugin in `sos`. This tool is delivered
+with RHEL, and is capable of generating reports for many applications in the
+system.
+
+Executing `sos report` yields information about a system and its components by
+using plugins to read each one of them. Plugins have an enablement check or may
+be invoked manually when executing the command.
+A [plugin](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin) is a
+python script that we include as part of `sos`.
+
+There is a [plugin for OpenShift](https://github.com/sosreport/sos/blob/main/sos/report/plugins/openshift.py)
+already that almost fits the needs of this EP. MicroShift does not share the
+same resources, APIs and namespaces/names that the plugin assumes, as it was
+made for OpenShift clusters. However, much of the needs covered in the plugin
+are the same, so the contents of the plugin may be used too.
+
+Apart from other components, that will be handled by their own plugins (full
+list in a system can be obtained by running `sos report -l`), we need to
+retrieve:
+1. Control plane component logs: apiserver, etcd, kubelet, etc.
+1. Yaml output for all the manifests.
+1. Logs for every container.
+1. MicroShift's configuration and version.
+
+Retrieving all this information involves using a kubernetes client. While
+MicroShift itself uses the go client extensively, the idea of introducing
+another command to produce this output and avoid installing `oc`/`kubectl` is
+not the best, since we would be adding runtime dependencies between a system
```

Comment:
As someone unfamiliar with microshift - would we not be able to use the go client that microshift uses itself? I.E. instead of the sos plugin running `oc get $thing`, is there something equivalent to `microshift get $thing` already available?

Reply:
The only available commands for microshift binary are:
```bash
$ microshift --help
MicroShift, a minimal OpenShift

Usage:
  microshift [flags]
  microshift [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  help        Help about any command
  run         Run MicroShift
  show-config Print MicroShift's configuration
  version     Print MicroShift version information

Flags:
  -h, --help      help for microshift
  -v, --v Level   number for the log level verbosity

Use "microshift [command] --help" for more information about a command.
```

What Frank is saying is that we can get anything from the cluster from within microshift binary, its already full of clients to the API server and we would dodge introducing `oc` for that.
My only concern, as mentioned in another comment, is having a generic-enough output format so that we dont end up changing the plugin every time something is added to `microshift must-gather`.

Reply:
Ah, understood. Re: doing this in the plugin via `oc` vs. adding a `microshift` command that collects everything itself and just calling that in the plugin, a potential additional wrench in a `microshift must-gather` or similar command that I just thought of...

One reason, beyond the uniformity to customers I mentioned earlier, for why we've ported several such tools from other projects to plugins is that with `--clean`, we remove any binary files by default as we can't guarantee sensitive information isn't included in them. So anything that produces a tarball or similar for sos to incorporate into the final archive would have that tarball removed if a security-minded customer ran `sos report --clean` for microshift.

Reply:
What I had in mind is a bit different.
Maybe have `microshift must-gather` return a json or some machine readable, structured format. Keys would determine a hierarchy for a file system, having dirs and files, and then the contents. Then the plugin would ingest the json and create the files under the `microshift` dir, following the same hierarchy.
Is that too much of an anti-pattern for `sos`?

Reply:
If we use oc now, we could add a `microshift get` later and change the sos plugin to be smart enough to decide which command to use.

In terms of a `microshift must-gather`, I understand the concerns with incorporating a binary blob. There's not that much inside the cluster that's unique for MicroShift, and the plugin currently lists a bunch of types and namespaces that should all be fetched. That seems fairly future-proof to me.

Reply:
It turns out we're already vendoring `kubectl get` so we just need to expose it. See https://github.com/openshift/microshift/pull/1280

Reply:
> Is that too much of an anti-pattern for sos?

Not an anti-pattern, but one we haven't used before. We did make some plugin API changes not too long ago that would support this kind of approach.

Reply:
Turns out the `microshift get` command is enough for everything, so reworking my implementation PR around it.

Reply:
Hey, just keep in mind that sosreport openshift uses the `oc` fully, so if the `microshift get` would not have the same logic as `oc`, the sosreport will fail.

Reply:
Turns out it requires an entire new plugin, as there are differences in how and what it retrieves.